base/inference.jl

```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
 
     if istopfunction(tm, f, :promote_type) || istopfunction(tm, f, :typejoin)
         return Type
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
```

Comment:
It's not just pure though, it also maps the type equality equivalence class to one value.